File: Sliding Window/1004. Max Consecutive Ones III/1004. Max Consecutive Ones III.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function (nums, k) {
  let longest = 0;
  let num_zeros = 0;
  let l = 0;

  for (let r = 0; r < nums.length; r++) {
    if (nums[r] === 0) num_zeros++;

    while (num_zeros > k) {
      if (nums[l] === 0) num_zeros--;
      l++;
    }

    longest = Math.max(longest, r - l + 1);
  }

  return longest;
};

// Time: O(n)
// Space: O(1)

File: Sliding Window/1004. Max Consecutive Ones III/1004. Max Consecutive Ones III.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        num_zeros = 0
        longest = 0
        
        l = 0
        for r in range(len(nums)):
            if nums[r] == 0:
                num_zeros += 1
            
            while num_zeros > k:
                if nums[l] == 0:
                    num_zeros -= 1
                l += 1

            longest = max(longest, r - l + 1)
        
        return longest

# Time: O(n)
# Space: O(1)